src/atomic/atomic_cell.rs

```diff
@@ -0,0 +1,862 @@
+            drop(self.replace(val));
+        } else {
+            unsafe {
+                atomic_store(self.value.get(), val);
+            }
+        }
+    }
+
+    /// Stores `val` into the atomic cell and returns the previous value.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_utils::atomic::AtomicCell;
+    ///
+    /// let a = AtomicCell::new(7);
+    ///
+    /// assert_eq!(a.get(), 7);
+    /// assert_eq!(a.replace(8), 7);
+    /// assert_eq!(a.get(), 8);
+    /// ```
+    pub fn replace(&self, val: T) -> T {
+        unsafe { atomic_swap(self.value.get(), val) }
+    }
+}
+
+impl<T: Default> AtomicCell<T> {
+    /// Takes the inner value and replaces it with `T::default()`.
+    ///
+    /// Note that `atomic_cell.take()` is equivalent to:
+    ///
+    /// ```ignore
+    /// atomic_cell.replace(T::default())
+    /// ```
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_utils::atomic::AtomicCell;
+    ///
+    /// let a = AtomicCell::new(7);
+    ///
+    /// assert_eq!(a.get(), 7);
+    /// assert_eq!(a.take(), 7);
+    /// assert_eq!(a.get(), 0);
+    /// ```
+    pub fn take(&self) -> T {
+        self.replace(T::default())
+    }
+}
+
+impl<T: Copy> AtomicCell<T> {
+    /// Returns a copy of the inner value.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_utils::atomic::AtomicCell;
+    ///
+    /// let a = AtomicCell::new(7);
+    ///
+    /// assert_eq!(a.get(), 7);
+    /// ```
+    pub fn get(&self) -> T {
+        unsafe { atomic_load(self.value.get()) }
+    }
+}
+
+impl<T: Copy + Eq> AtomicCell<T> {
+    /// If the current value equals `current`, stores `new` into the atomic cell.
+    ///
+    /// Returns `true` if the value was updated, and `false` otherwise.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use crossbeam_utils::atomic::AtomicCell;
+    ///
+    /// #[derive(Copy, Clone, Eq)]
+    /// struct Foo(i32);
+    ///
+    /// impl PartialEq for Foo {
+    ///     fn eq(&self, other: &Foo) -> bool {
+    ///         self.0 % 10 == other.0 % 10
+    ///     }
+    /// }
+    ///
+    /// let a = AtomicCell::new(Foo(1));
+    ///
+    /// assert_eq!(a.compare_and_set(Foo(2), Foo(3)), false);
+    /// assert_eq!(a.get().0, 1);
+    ///
+    /// assert_eq!(a.compare_and_set(Foo(11), Foo(12)), true);
+    /// assert_eq!(a.get().0, 12);
+    /// ```
+    pub fn compare_and_set(&self, mut current: T, new: T) -> bool {
+        loop {
+            match unsafe { atomic_compare_exchange_weak(self.value.get(), current, new) } {
+                Ok(_) => return true,
+                Err(previous) => {
+                    if previous != current {
+                        return false;
+                    }
+
+                    // The compare-exchange operation has failed and didn't store `new`. However,
+                    // `previous` is semantically equal to `current`, although not byte-equal.
```

Comment:
They can also be byte-equal, as we're using the weak version of CAS here.

Reply:
True. I've updated the comment.